### PR TITLE
dns-domain: fix build failure with libidn

### LINKED ISSUE #19723: systemd fails to compile with `libidn`
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > CentOS 7

```
Run-time dependency libidn2 found: NO (tried pkgconfig and cmake)
Run-time dependency libidn found: YES 1.28
...
../src/shared/dns-domain.c: In function ‘dns_name_apply_idna’:
../src/shared/dns-domain.c:1360:87: error: macro "GREEDY_REALLOC" passed 3 arguments, but takes just 2
                 if (!GREEDY_REALLOC(buf, allocated, n + !first + DNS_LABEL_ESCAPED_MAX))
                                                                                       ^
../src/shared/dns-domain.c:1360:22: error: ‘GREEDY_REALLOC’ undeclared (first use in this function)
                 if (!GREEDY_REALLOC(buf, allocated, n + !first + DNS_LABEL_ESCAPED_MAX))
                      ^
../src/shared/dns-domain.c:1360:22: note: each undeclared identifier is reported only once for each function it appears in
../src/shared/dns-domain.c:1378:50: error: macro "GREEDY_REALLOC" passed 3 arguments, but takes just 2
         if (!GREEDY_REALLOC(buf, allocated, n + 1))
                                                  ^
../src/shared/dns-domain.c:1338:23: error: unused variable ‘allocated’ [-Werror=unused-variable]
         size_t n = 0, allocated = 0;
                       ^
cc1: all warnings being treated as errors
```

I guess the `HAVE_LIBIDN` guarded code in `src/shared/dns-domain.c` needs to be updated to match the `GREEDY_REALLOC()` expectations from 319a4f4bc46b230fc660321e99aaac1bc449deea.

/cc @poettering

### LINKED ISSUE #19725: systemd crashes when reexecuting from v219 to v249
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > CentOS 7

When trying to build & install latest master on CentOS 7, systemd crashes right after running `systemctl daemon-reexec`:

```
[root@n56 ~]# systemctl daemon-reexec
[root@n56 ~]# 
Message from syslogd@n56 at May 25 20:46:16 ...
 systemd[1]:Caught <ABRT>, dumped core as pid 17243.

Message from syslogd@n56 at May 25 20:46:16 ...
 systemd[1]:Freezing execution.
```

```
[ 1062.748130] systemd[1]: systemd 248-1202-g9caf985 running in system mode. (+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS -OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 -IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP -LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB -ZSTD +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
[ 1062.068428] systemd[1]: Unwatching 945.
[ 1062.779630] systemd[1]: Detected architecture x86-64.
[ 1062.098373] systemd[1]: Unwatching 943.
[ 1062.099210] systemd[1]: Unwatching 709.
[ 1062.099842] systemd[1]: Unwatching 9890.
[ 1062.103116] systemd[1]: Unwatching 1017.
[ 1062.109349] systemd[1]: Unwatching 996.
[ 1062.109857] systemd[1]: Unwatching 952.
[ 1062.110713] systemd[1]: Unwatching 993.
[ 1062.116026] systemd[1]: Unwatching 1032.
[ 1062.123011] systemd[1]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=255 reply_cookie=0 error=n/a
[ 1062.123178] systemd[1]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=8996 reply_cookie=0 error=n/a
[ 1062.123350] systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=8997 reply_cookie=0 error=n/a
[ 1062.806870] systemd-rc-local-generator[17237]: /etc/rc.local is not marked executable, skipping.
[ 1062.808111] systemd-sysv-generator[17240]: SysV service '/etc/rc.d/init.d/netconsole' lacks a native systemd unit file. Automatically generating a unit file for compatibility. Please update package to include a native systemd unit file, in order to make it more safe and robust.
[ 1062.808161] systemd-sysv-generator[17240]: SysV service '/etc/rc.d/init.d/network' lacks a native systemd unit file. Automatically generating a unit file for compatibility. Please update package to include a native systemd unit file, in order to make it more safe and robust.
[ 1062.910421] systemd[1]: /usr/lib/systemd/system/dracut-pre-udev.service:31: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1062.930664] systemd[1]: /usr/lib/systemd/system/dracut-pre-udev.service:32: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1062.953191] systemd[1]: /usr/lib/systemd/system/dracut-pre-trigger.service:28: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1062.973705] systemd[1]: /usr/lib/systemd/system/dracut-pre-trigger.service:29: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1062.996397] systemd[1]: /usr/lib/systemd/system/dracut-pre-pivot.service:27: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1063.016738] systemd[1]: /usr/lib/systemd/system/dracut-pre-pivot.service:28: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1063.041607] systemd[1]: /usr/lib/systemd/system/dracut-pre-mount.service:28: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1063.061943] systemd[1]: /usr/lib/systemd/system/dracut-pre-mount.service:29: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1063.084257] systemd[1]: /usr/lib/systemd/system/dracut-mount.service:26: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1063.104277] systemd[1]: /usr/lib/systemd/system/dracut-mount.service:27: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1063.126130] systemd[1]: /usr/lib/systemd/system/dracut-initqueue.service:34: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1063.146489] systemd[1]: /usr/lib/systemd/system/dracut-initqueue.service:35: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1063.168631] systemd[1]: /usr/lib/systemd/system/dracut-cmdline.service:31: Standard output type syslog is obsolete, automatically updating to journal. Please update your unit file, and consider removing the setting altogether.
[ 1063.188777] systemd[1]: /usr/lib/systemd/system/dracut-cmdline.service:32: Standard output type syslog+console is obsolete, automatically updating to journal+console. Please update your unit file, and consider removing the setting altogether.
[ 1063.225492] systemd[1]: /usr/lib/systemd/system/rpcbind.socket:5: ListenStream= references a path below legacy directory /var/run/, updating /var/run/rpcbind.sock → /run/rpcbind.sock; please update the unit file accordingly.
[ 1063.250819] systemd[1]: /usr/lib/systemd/system/chronyd.service:10: PIDFile= references a path below legacy directory /var/run/, updating /var/run/chrony/chronyd.pid → /run/chrony/chronyd.pid; please update the unit file accordingly.
[ 1063.295135] systemd[1]: /usr/lib/systemd/system/plymouth-start.service:13: Unit configured to use KillMode=none. This is unsafe, as it disables systemd's process lifecycle management for the service. Please update your service to use a safer KillMode=, such as 'mixed' or 'control-group'. Support for KillMode=none is deprecated and will eventually be removed.
[ 1063.384982] systemd[1]: Assertion '(_error) != 0' failed at src/core/service.c:2940, function service_deserialize_item(). Aborting.
[ 1063.401410] systemd[1]: Caught <ABRT>, dumped core as pid 17243.
[ 1063.407450] systemd[1]: Freezing execution.
```

### PR DESCRIPTION
**Pull request by yuwata**
Follow-up for 319a4f4bc46b230fc660321e99aaac1bc449deea.

Fixes #19723.
Fixes #19725.

### CODE REVIEW

**Comment by yuwata** [comment]
cc @mrc0mmand

I have not tested this patch...

**Comment by mrc0mmand** [comment]
Thank you! I can confirm the build now finishes successfully. I can't test it further due to #19725, but the build issue is definitely resolved.

**Comment by yuwata** [comment]
@mrc0mmand One more commit is added to fix #19725.

File: src/core/service.c [hunk 1]
```diff
@@ -2937,7 +2937,7 @@ static int service_deserialize_item(Unit *u, const char *key, const char *value,
 
                 r = extract_first_word(&value, &fdn, NULL, EXTRACT_CUNESCAPE | EXTRACT_UNQUOTE);
                 if (r <= 0) {
-                        log_unit_debug_errno(u, r, "Failed to parse fd-store-fd value \"%s\": %m", value);
+                        log_unit_debug(u, "Failed to parse fd-store-fd value: %s", value);
```
Comment (poettering) [inline src/core/service.c:2940]: needs zerook, no?

**Comment by mrc0mmand** [comment]
> @mrc0mmand One more commit is added to fix #19725.

Wow, that was fast. And I can confirm that both of the fixes work, the daemon reexecutes correctly, thanks!

File: src/core/service.c (hunk 1 above)
Comment (yuwata) [reply src/core/service.c:2940]: Hmm, `Failed to parse fd-store-fd value "123": Success` sounds strange. Another option is to call `log_unit_debug_errno()` for `r < 0` and `log_unit_debug()` for `r == 0`. But I think that is mostly redundant...

Comment (poettering) [reply src/core/service.c:2940]: true

**Comment by poettering** [comment]
lgtm, too

**Pull request merged**